hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineManagerV2Impl.java

```diff
@@ -0,0 +1,637 @@
+        "SCMPipelineManagerInfo", this);
+    this.metrics = SCMPipelineMetrics.create();
+    this.pipelineWaitDefaultTimeout = conf.getTimeDuration(
+        HddsConfigKeys.HDDS_PIPELINE_REPORT_INTERVAL,
+        HddsConfigKeys.HDDS_PIPELINE_REPORT_INTERVAL_DEFAULT,
+        TimeUnit.MILLISECONDS);
+    this.isInSafeMode = new AtomicBoolean(conf.getBoolean(
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_ENABLED,
+        HddsConfigKeys.HDDS_SCM_SAFEMODE_ENABLED_DEFAULT));
+    // Pipeline creation is only allowed after the safemode prechecks have
+    // passed, eg sufficient nodes have registered.
+    this.pipelineCreationAllowed = new AtomicBoolean(!this.isInSafeMode.get());
+    initializePipelineState();
+  }
+
+  public static PipelineManager newPipelineManager(
+      ConfigurationSource conf, SCMHAManager scmhaManager,
+      NodeManager nodeManager, Table<PipelineID, Pipeline> pipelineStore,
+      PipelineFactory pipelineFactory) throws IOException {
+    // Create PipelineStateManager
+    PipelineStateManagerV2 stateManager = PipelineStateManagerV2Impl
+        .newBuilder().setPipelineStore(pipelineStore)
+        .setRatisServer(scmhaManager.getRatisServer()).build();
+
+    // Create PipelineManager
+    PipelineManagerV2Impl pipelineManager = new PipelineManagerV2Impl(conf,
+        nodeManager, stateManager, pipelineFactory);
+
+    // Create background thread.
+    Scheduler scheduler = new Scheduler(
+        "RatisPipelineUtilsThread", false, 1);
+    BackgroundPipelineCreator backgroundPipelineCreator =
+        new BackgroundPipelineCreator(pipelineManager, scheduler, conf);
+    pipelineManager.setBackgroundPipelineCreator(backgroundPipelineCreator);
+    pipelineManager.setScheduler(scheduler);
+
+    return pipelineManager;
+  }
+
+  protected void initializePipelineState() throws IOException {
+    if (stateManager.isPipelineStoreEmpty()) {
+      LOG.info("No pipeline exists in current db");
+      return;
+    }
+    TableIterator<PipelineID, ? extends Table.KeyValue<PipelineID, Pipeline>>
+        iterator = stateManager.getPipelineStoreIterator();
+    while (iterator.hasNext()) {
+      Pipeline pipeline = iterator.next().getValue();
+      stateManager.addPipeline(pipeline.getProtobufMessage());
+      nodeManager.addPipeline(pipeline);
+    }
+  }
+
+  @Override
+  public Pipeline createPipeline(ReplicationType type,
+                                 ReplicationFactor factor) throws IOException {
+    if (!isPipelineCreationAllowed() && factor != ReplicationFactor.ONE) {
+      LOG.debug("Pipeline creation is not allowed until safe mode prechecks " +
+          "complete");
+      throw new IOException("Pipeline creation is not allowed as safe mode " +
+          "prechecks have not yet passed");
+    }
+    lock.writeLock().lock();
+    try {
+      Pipeline pipeline = pipelineFactory.create(type, factor);
+      stateManager.addPipeline(pipeline.getProtobufMessage());
+      nodeManager.addPipeline(pipeline);
```

Comment:
`stateManager#addPipeline` is replicated via Ratis, which will make the pipeline added to in-memory and to DB on all the SCMs.
`nodeManager#addPipeline`, since executed outside of `PipelineStateManager` will only be executed on Leader SCM. This will be a problem as the followers will not have this information.

Move `nodeManager#addPipeline` call to `stateManager#addPipeline`.

Reply:
OK Will do